### PR TITLE
Implement transport circuit breaking in aggregator

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit moves the action name validation and circuit breaking into
the InboundAggregator. This work is valuable because it lays the 
groundwork for incrementally circuit breaking as data is received.

This PR includes the follow behavioral change:

Handshakes contribute to circuit breaking, but cannot be broken. They
currently do not contribute nor are they broken.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 1]
```diff
@@ -147,54 +144,59 @@ private void messageReceived(InboundMessage message, TcpChannel channel) throws
...
-                logger.warn(() -> new ParameterizedMessage("Failed to send error message back to client for action [{}]", action), inner);
+                sendErrorResponse(action, transportChannel, e);
             }
         }
     }
 
+    private static void sendErrorResponse(String actionName, TransportChannel transportChannel, Exception e) {
+        try {
+            transportChannel.sendResponse(e);
+        } catch (IOException inner) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:194]: should we catch `Exception` here? We probably never want to bubble anything up here

File: server/src/main/java/org/elasticsearch/transport/InboundAggregator.java [hunk 2]
```diff
@@ -108,23 +156,84 @@ public void close() {
...
+            this.circuitBreaker = circuitBreaker;
+        }
+
+        private void incrementReservedBytes(int delta) {
+            bytesToRelease.getAndAdd(delta);
+        }
+
+        @Override
+        public void close() {
+            final int toRelease = bytesToRelease.getAndSet(0);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/InboundAggregator.java:235]: I think we should protect against double-closing here, given how important it is to this correctly.

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java [hunk 3]
```diff
@@ -74,6 +75,18 @@ public void close() {
     }
 
     public void handleBytes(TcpChannel channel, ReleasableBytesReference reference) throws IOException {
+        if (uncaughtException != null) {
+            throw new IllegalStateException("Pipeline state corrupted by uncaught exception", uncaughtException);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:79]: when do we expect this to happen? should we `assert false` here?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks Tim. I've left some comments.

File: server/src/main/java/org/elasticsearch/transport/InboundAggregator.java (hunk 2 above)
Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/transport/InboundAggregator.java:235]: I made a change.

File: server/src/main/java/org/elasticsearch/transport/InboundAggregator.java [hunk 4]
```diff
@@ -108,23 +156,84 @@ public void close() {
...
+
+        private final Supplier<CircuitBreaker> circuitBreaker;
+        private final AtomicInteger bytesToRelease = new AtomicInteger(0);
+
+        private BreakerControl(Supplier<CircuitBreaker> circuitBreaker) {
+            this.circuitBreaker = circuitBreaker;
+        }
+
+        private void incrementReservedBytes(int delta) {
+            bytesToRelease.getAndAdd(delta);
```
[the remark this answers is not in the document]
Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/transport/InboundAggregator.java:230]: I made a change and renamed the method to `set`.

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java (hunk 1 above)
Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/transport/InboundHandler.java]: I did a little more clean-up around here. I still think we need a follow-up dedicated to exception handling. But I tried to maintain consistent behavior while moving the correct direction.

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 5]
```diff
@@ -147,54 +145,57 @@ private void messageReceived(InboundMessage message, TcpChannel channel) throws
...
-                transportChannel = new TcpTransportChannel(outboundHandler, channel, action, requestId, version,
-                    circuitBreakerService, 0, header.isCompressed(), header.isHandshake());
+            // Cannot short circuit handshakes
+            assert message.isShortCircuit() == false;
+            final StreamInput stream = namedWriteableStream(message.openOrGetStreamInput());
+            assertRemoteVersion(stream, header.getVersion());
+            try (Releasable breakerRelease = message.takeBreakerReleaseControl()) {
+                final TransportChannel transportChannel = new TcpTransportChannel(outboundHandler, channel, action, requestId, version,
+                    header.isCompressed(), header.isHandshake(), () -> {});
                 handshaker.handleHandshake(transportChannel, requestId, stream);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:161]: If `TransportHandshaker.handleHandshake` throws an exception (e.g. IllegalStateException), that is no longer bubbled up back to the node that initiated the handshake. I'm not sure what this change of behavior entails, but would suggest backing it out of this PR.

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left one more comment, looking good o.w.

Would appreciate a second pair of eyes from @original-brownbear, as this is critical infrastructure code.

**Review by original-brownbear — COMMENTED** [review]
I think the code is ok, but I'm a little uneasy about the plan here:

>  incrementally circuit breaking as data is received

Are we sure this is what we want in the first place? It seems to me what we really want is to circuit break before reading full messages.
If we circuit break incrementally, that means we're blowing through a bunch of buffer space only to then throw away a message mid-way. How would we ensure liveness here? If a node has a number of large messages come in concurrently, it will never process any of them but waste buffers for all of them for large (smaller than available memory) message sizes given enough concurrent messages?

Wouldn't we instead want to check the circuit breaker in `headerReceived`? That way we can circuit break a message without wasting all the buffers for it. We can increment the circuit breaker right after reading the header. 
If we trip it, we just drop/release all the bytes in the remainder of the message as we aggregate and prevent needlessly holding on to buffered bytes that will never be deserialized.
If we don't trip it, we would've incremented it already before reading  the message and would have the guarantee that the buffers we starting holding on to for it, will not be wasted?

=> to me it seems like that's what we want and could have in this PR at low cost with the changes it introduces.

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java (hunk 5 above)
Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/transport/InboundHandler.java:160]: Made this change.

**Comment by tbrooks8** [comment]
> Are we sure this is what we want in the first place?

This commit does not change any circuit breaking behavior. The end result of what circuit breaking will look like I imagine will be a combination of incremental and pre-breaking bytes that we know we are about to received.

The header int will inform this work. But it is not always present, accurate (compression), and there is some tension between that and our current usage of the MXBeans heap stats to circuit break. So the end result I think will look something like what you're describing. But, there are some complications and this is only an infrastructure PR.

**Review by original-brownbear — APPROVED** [review]
LGTM, thanks for the explanations Tim! Definitely more flexible to have the circuit breaking further up stream no matter what we do going forward :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**